### ISSUE TITLE
"where" is not filtering (when combine with each and another table)

### ISSUE THREAD

**Issue by wdv4758h** [nushell/nushell#9271]
### Describe the bug

I think this `where` is wierd. It's not doing any filtering. Let me know if it's my misunderstanding.

### How to reproduce

```nu
let x = [
  {a: 1, b: 2}
  {a: 2, b: 3}
  {a: 3, b: 4}
]

let y = [
  {a: 1, b: 5}
  {a: 2, b: 6}
  {a: 3, b: 7}
]

$x | each {|it| $y | where a == $it.a}

# current output:
# ╭───┬───────────────╮
# │ 0 │ ╭───┬───┬───╮ │
# │   │ │ # │ a │ b │ │
# │   │ ├───┼───┼───┤ │
# │   │ │ 0 │ 1 │ 5 │ │
# │   │ │ 1 │ 2 │ 6 │ │
# │   │ │ 2 │ 3 │ 7 │ │
# │   │ ╰───┴───┴───╯ │
# │ 1 │ ╭───┬───┬───╮ │
# │   │ │ # │ a │ b │ │
# │   │ ├───┼───┼───┤ │
# │   │ │ 0 │ 1 │ 5 │ │
# │   │ │ 1 │ 2 │ 6 │ │
# │   │ │ 2 │ 3 │ 7 │ │
# │   │ ╰───┴───┴───╯ │
# │ 2 │ ╭───┬───┬───╮ │
# │   │ │ # │ a │ b │ │
# │   │ ├───┼───┼───┤ │
# │   │ │ 0 │ 1 │ 5 │ │
# │   │ │ 1 │ 2 │ 6 │ │
# │   │ │ 2 │ 3 │ 7 │ │
# │   │ ╰───┴───┴───╯ │
# ╰───┴───────────────╯
```

### Expected behavior

That where condition should work like this `filter`:

```nu
let x = [
  {a: 1, b: 2}
  {a: 2, b: 3}
  {a: 3, b: 4}
]

let y = [
  {a: 1, b: 5}
  {a: 2, b: 6}
  {a: 3, b: 7}
]

$x | each {|it| $y | filter { $in.a == $it.a} }

# output:
# ╭───┬───────────────╮
# │ 0 │ ╭───┬───┬───╮ │
# │   │ │ # │ a │ b │ │
# │   │ ├───┼───┼───┤ │
# │   │ │ 0 │ 1 │ 5 │ │
# │   │ ╰───┴───┴───╯ │
# │ 1 │ ╭───┬───┬───╮ │
# │   │ │ # │ a │ b │ │
# │   │ ├───┼───┼───┤ │
# │   │ │ 0 │ 2 │ 6 │ │
# │   │ ╰───┴───┴───╯ │
# │ 2 │ ╭───┬───┬───╮ │
# │   │ │ # │ a │ b │ │
# │   │ ├───┼───┼───┤ │
# │   │ │ 0 │ 3 │ 7 │ │
# │   │ ╰───┴───┴───╯ │
# ╰───┴───────────────╯
```

### Screenshots

![image](https://github.com/nushell/nushell/assets/2716047/e9c6368f-ad8b-4887-b75c-0ade3fd7cd76)

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.80.1                                   |
| branch             | main                                     |
| commit_hash        | 74724dee809c039440ce299db65ee5ad5aa321ea |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.67.1 (d5a82bbd2 2023-02-07)      |
| rust_channel       | 1.67.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.67.1 (8ecd4f20a 2023-01-10)      |
| build_time         | 2023-05-24 15:51:37 +08:00               |
| build_rust_channel | release                                  |
| features           | default, sqlite, trash, which, zip       |
| installed_plugins  | query, query json, query web, query xml  |

**Comment by wdv4758h** [comment]
I find rename that var of each works. Calling it `it` will not work. But calling it like `r` will work.

![image](https://github.com/nushell/nushell/assets/2716047/20ff72f3-7867-4183-ad77-bb9e48b75f38)

**Comment by wdv4758h** [comment]
I guess `it` is a special name in this case?

**Comment by amtoine** [comment]
i'm not sure what the expected output here is :open_mouth: 

but the fact that, on the same `x` and `y`,
```
$x | each {|it| $y | where a == $it.a}
```
and
```
$x | each {|it| $y | filter { $in.a == $it.a} }
```
do not give the same output sounds wrong to me :thinking:

**Comment by weirdan** [comment]
`$it` is a special name used in `where` to reference the current element, so `a == $it.a` is always true (it refers to the same value).

**Comment by WindSoilder** [comment]
Yeah because `$it` is special in `where`, I think it's good to use other variable name in `where`.

I'd like to close the issue, feel free to comment or re-open it if you don't agree
